File: test/language/templates/cpp98/StructFullNameTemplateArgumentTest.cpp
#include "gtest/gtest.h"

#include "templates/struct_full_name_template_argument/StructFullNameTemplateArgument.h"

namespace templates
{
namespace struct_full_name_template_argument
{

TEST(StructFullNameTemplateArgumentTest, readWrite)
{
    StructFullNameTemplateArgument structFullNameTemplateArgument;
    TemplatedStruct_Storage_08C0ED6D& externalStruct = structFullNameTemplateArgument.getStructExternal();
    externalStruct.getStorage().setData(42);
    TemplatedStruct_Storage_A3A4B101& internalStruct = structFullNameTemplateArgument.getStructInternal();
    internalStruct.getStorage().setData("string");

    zserio::BitStreamWriter writer;
    structFullNameTemplateArgument.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    StructFullNameTemplateArgument readStructFullNameTemplateArgument(reader);

    ASSERT_TRUE(structFullNameTemplateArgument == readStructFullNameTemplateArgument);
}

} // namespace struct_full_name_template_argument
} // namespace templates

File: test/extensions/language/templates/java/templates/StructFullNameTemplateArgumentTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.struct_full_name_template_argument.Storage;
import templates.struct_full_name_template_argument.StructFullNameTemplateArgument;
import templates.struct_full_name_template_argument.TemplatedStruct_Storage_A3A4B101;
import templates.struct_full_name_template_argument.TemplatedStruct_Storage_C76E422F;

public class StructFullNameTemplateArgumentTest
{
    @Test
    public void readWrite() throws IOException
    {
        final StructFullNameTemplateArgument structFullNameTemplateArgument =
                new StructFullNameTemplateArgument();
        structFullNameTemplateArgument.setStructExternal(new TemplatedStruct_Storage_C76E422F(
                new templates.struct_full_name_template_argument.import_storage.Storage((long)42)));
        structFullNameTemplateArgument.setStructInternal(
                new TemplatedStruct_Storage_A3A4B101(new Storage("string")));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        structFullNameTemplateArgument.write(writer);

        final BitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final StructFullNameTemplateArgument readStructFullNameTemplateArgument =
                new StructFullNameTemplateArgument(reader);
        assertTrue(structFullNameTemplateArgument.equals(readStructFullNameTemplateArgument));
    }
}
